# Data inspection & exploration: STARR-seq CD4 T cells

Paper: Cis-regulatory atlas of primary human CD4+ T cells - Kurtis Stefan & Artem Barski 2023

Contains preparation for grch build conversion of the enhancer and NRE regions of CD4+ T-cells of lentistarr MPRA<br>
Saves the correct information/columns to txt file in BED format

### Import libraries

In [ ]:
import pandas as pd
import numpy as np
from read_config_file import get_config
import re

In [181]:
pd.set_option('mode.chained_assignment', None)

- tpm: transcripts per million

Enhancers are called with MACS2 [-log10(P) > 75], comparing STARR-Seq–obtained RNA reads to input plasmid reads.<br>
Negative regulatory elements (NREs) are called using MACS2 [-log10(P) > 30] comparing input plasmid to STARR-Seq-obtained RNA reads

### Load data

In [182]:
config = get_config()
starr_cd4_enh = (config['starr_cd4_enhancer'])
starr_cd4_nre = (config['starr_cd4_nre'])

In [183]:
enh = pd.read_csv(starr_cd4_enh, sep=';')

In [184]:
nre = pd.read_csv(starr_cd4_nre, sep=';')

In [185]:
lcl_analyis = pd.read_csv('positions_sig_lcl_variants.csv')

- the final analysis data are not strand specific.
- Genome build: hg19 - grch37

### Data inspection, explorationa and cleaning

In [186]:
enh

chromosome  start_coord  end_coord  \
0          chr1       780062     780170   
1          chr1      1051480    1051842   
2          chr1      1136882    1137106   
3          chr1      1143458    1143553   
4          chr1      1167340    1167536   
...         ...          ...        ...   
8409       chrY     21906762   21906867   
8410       chrY     22737445   22737756   
8411       chrY     23167461   23167667   
8412       chrY     23402278   23402444   
8413       chrY     23402586   23402698   

                                                   name  score strand  \
0     MACS2STARRENH_indivuallyrmDupat75thpercentile0...   2990      .   
1     MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1773      .   
2     MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1835      .   
3     MACS2STARRENH_indivuallyrmDupat75thpercentile0...   2820      .   
4     MACS2STARRENH_indivuallyrmDupat75thpercentile0...   4380      .   
...                                                 ...    ...    ...   
8409  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1514      .   
8410  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1109      .   
8411  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1066      .   
8412  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   3543      .   
8413  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1311      .   

      signalValue  pvalue(-log10)  qvalue(-log10)  peak  
0         6.76315       302.37964       299.05209    42  
1         1.81547       180.31976       177.37407   121  
2         3.38474       186.49478       183.52959    97  
3         2.61547       285.31268       282.04059    72  
4         2.54028       441.82684       438.00470   138  
...           ...             ...             ...   ...  
8409      3.45702       154.32509       151.46123    76  
8410      1.79933       113.70029       110.97779   202  
8411     10.88384       109.34882       106.64301   168  
8412      3.65531       357.82581       354.30984    92  
8413      3.30078       133.90356       131.10869    55  

[8414 rows x 10 columns]

In [187]:
enh.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8414 entries, 0 to 8413
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   chromosome      8414 non-null   object 
 1   start_coord     8414 non-null   int64  
 2   end_coord       8414 non-null   int64  
 3   name            8414 non-null   object 
 4   score           8414 non-null   int64  
 5   strand          8414 non-null   object 
 6   signalValue     8414 non-null   float64
 7   pvalue(-log10)  8414 non-null   float64
 8   qvalue(-log10)  8414 non-null   float64
 9   peak            8414 non-null   int64  
dtypes: float64(3), int64(4), object(3)
memory usage: 657.5+ KB


In [188]:
enh.isna().sum()

chromosome        0
start_coord       0
end_coord         0
name              0
score             0
strand            0
signalValue       0
pvalue(-log10)    0
qvalue(-log10)    0
peak              0
dtype: int64

In [189]:
enh[['score', 'signalValue', 'peak']].describe()

score  signalValue         peak
count   8414.000000  8414.000000  8414.000000
mean    1914.187307     3.223809   112.933801
std     1286.327638     2.521876    77.818372
min      724.000000     1.265230     0.000000
25%     1046.000000     1.950638    59.000000
50%     1514.000000     2.535415    92.000000
75%     2341.000000     3.608805   149.000000
max    16172.000000    75.354660   771.000000

In [190]:
enh['name'].value_counts()

MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_127916    2
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_134214    2
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_157104    2
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_148240    2
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_128980    1
                                                                  ..
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_66654     1
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_66645     1
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_66644     1
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_66641     1
MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_211382    1
Name: name, Length: 8410, dtype: int64

In [191]:
#Show the peaks with duplicate names
enh[enh['name'].map(enh['name'].value_counts() > 1)]

chromosome  start_coord  end_coord  \
5570       chr3     98282557   98282750   
5571       chr3     98283019   98283148   
5769       chr3    179370385  179370561   
5770       chr3    179370992  179371042   
6133       chr4    183838469  183838571   
6134       chr4    183838737  183838875   
6367       chr5    122181094  122181177   
6368       chr5    122181336  122181362   

                                                   name  score strand  \
5570  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1021      .   
5571  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1021      .   
5769  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   2277      .   
5770  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   2277      .   
6133  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1282      .   
6134  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   1282      .   
6367  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   2280      .   
6368  MACS2STARRENH_indivuallyrmDupat75thpercentile0...   2280      .   

      signalValue  pvalue(-log10)  qvalue(-log10)  peak  
5570      5.66464       104.83451       102.14629   215  
5571      5.66464       104.83451       102.14629   215  
5769      4.53904       230.88206       227.77855   144  
5770      4.53904       230.88206       227.77855   144  
6133      2.70363       131.08398       128.29875    63  
6134      2.70363       131.08398       128.29875    63  
6367      3.00476       231.12044       228.01610    33  
6368      3.00476       231.12044       228.01610    33

For the some peaks there are multiple genomic positions. However they have the same values in the rows.

In [192]:
enh.strand.value_counts()

.    8414
Name: strand, dtype: int64

In [193]:
# Calculate the qvalue
enh['qvalue'] = enh['qvalue(-log10)'].map(lambda x: 10 ** -x)

In [194]:
#Drop strand column and rearrange columns
enh = enh[[ 'name','chromosome', 'start_coord', 'end_coord', 'score', 'signalValue', 'pvalue(-log10)', 'qvalue(-log10)', 'qvalue', 'peak']]

In [195]:
enh[enh['qvalue'] <= 0.05]

name chromosome  \
0     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
1     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
2     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
3     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
4     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
...                                                 ...        ...   
8409  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chrY   
8410  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chrY   
8411  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chrY   
8412  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chrY   
8413  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chrY   

      start_coord  end_coord  score  signalValue  pvalue(-log10)  \
0          780062     780170   2990      6.76315       302.37964   
1         1051480    1051842   1773      1.81547       180.31976   
2         1136882    1137106   1835      3.38474       186.49478   
3         1143458    1143553   2820      2.61547       285.31268   
4         1167340    1167536   4380      2.54028       441.82684   
...           ...        ...    ...          ...             ...   
8409     21906762   21906867   1514      3.45702       154.32509   
8410     22737445   22737756   1109      1.79933       113.70029   
8411     23167461   23167667   1066     10.88384       109.34882   
8412     23402278   23402444   3543      3.65531       357.82581   
8413     23402586   23402698   1311      3.30078       133.90356   

      qvalue(-log10)         qvalue  peak  
0          299.05209  8.869722e-300    42  
1          177.37407  4.226005e-178   121  
2          183.52959  2.953997e-184    97  
3          282.04059  9.107727e-283    72  
4          438.00470   0.000000e+00   138  
...              ...            ...   ...  
8409       151.46123  3.457562e-152    76  
8410       110.97779  1.052471e-111   202  
8411       106.64301  2.275045e-107   168  
8412       354.30984   0.000000e+00    92  
8413       131.10869  7.785921e-132    55  

[8414 rows x 10 columns]

All rows are already below a qvalue of 0.05. Everything can be used for analysis.

In [196]:
#mpra_enh = enh[['chromosome', 'start_coord', 'end_coord', 'name']]

In [197]:
#mpra_enh.to_csv('data/mpra_enhancer_coords.txt', index=False, header=None, sep='\t')

In [198]:
print('There are ', len(enh[enh.peak == 0]), 'rows/peak that have a peaknr of 0')

There are  5 rows/peak that have a peaknr of 0


In [199]:
# There are is no eQTL data available for X and Y chromosome in the other dataset, so leave out the rows for chrX and chrY
enh = enh[~enh['chromosome'].isin(['chrX', 'chrY'])]
enh

name chromosome  \
0     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
1     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
2     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
3     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
4     MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr1   
...                                                 ...        ...   
8150  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr9   
8151  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr9   
8152  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr9   
8153  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr9   
8154  MACS2STARRENH_indivuallyrmDupat75thpercentile0...       chr9   

      start_coord  end_coord  score  signalValue  pvalue(-log10)  \
0          780062     780170   2990      6.76315       302.37964   
1         1051480    1051842   1773      1.81547       180.31976   
2         1136882    1137106   1835      3.38474       186.49478   
3         1143458    1143553   2820      2.61547       285.31268   
4         1167340    1167536   4380      2.54028       441.82684   
...           ...        ...    ...          ...             ...   
8150    140473478  140473662   2493      3.11855       252.54643   
8151    140485008  140485172   1632      2.20556       166.12692   
8152    140499611  140499764   1355      1.55885       138.31764   
8153    140523078  140523281   2098      5.49343       212.87029   
8154    140617195  140617347   2050      5.34030       208.10439   

      qvalue(-log10)         qvalue  peak  
0          299.05209  8.869722e-300    42  
1          177.37407  4.226005e-178   121  
2          183.52959  2.953997e-184    97  
3          282.04059  9.107727e-283    72  
4          438.00470   0.000000e+00   138  
...              ...            ...   ...  
8150       249.37758  4.191988e-250     6  
8151       163.22536  5.951686e-164    42  
8152       135.50819  3.103202e-136   131  
8153       209.82153  1.508238e-210    91  
8154       205.07066  8.498455e-206    80  

[8155 rows x 10 columns]

In [200]:
#enh.to_csv('data/cd4_enhancer_sequenties.csv', index=False, sep='\t')

### from build 37 to 38

Create a bedfile for conversion grom build 37 to build 38

In [201]:
#Select the relevant columns:
enh_bed = enh[['chromosome', 'start_coord', 'end_coord', 'name']]

In [202]:
enh_bed

chromosome  start_coord  end_coord  \
0          chr1       780062     780170   
1          chr1      1051480    1051842   
2          chr1      1136882    1137106   
3          chr1      1143458    1143553   
4          chr1      1167340    1167536   
...         ...          ...        ...   
8150       chr9    140473478  140473662   
8151       chr9    140485008  140485172   
8152       chr9    140499611  140499764   
8153       chr9    140523078  140523281   
8154       chr9    140617195  140617347   

                                                   name  
0     MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
1     MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
2     MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
3     MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
4     MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
...                                                 ...  
8150  MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
8151  MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
8152  MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
8153  MACS2STARRENH_indivuallyrmDupat75thpercentile0...  
8154  MACS2STARRENH_indivuallyrmDupat75thpercentile0...  

[8155 rows x 4 columns]

There are some duplicate names, but they have different positions. Adding the positions the original name to create unique names

In [203]:
#Set datatype of dataframe to string
enh_bed = enh_bed.astype(str)

In [204]:
#Create a unique name per row based on position
enh_bed['unique_name'] = enh_bed['chromosome'] +'_'+ enh_bed['start_coord'] +'_'+ enh_bed['end_coord']+ '_' + enh_bed['name']


In [205]:
#Select relevant columns with the new unique name
enh_bed = enh_bed[['chromosome', 'start_coord', 'end_coord', 'unique_name']]

In [206]:
#Write the columns to a txt file
enh_bed.to_csv('enhancer_coordinates.txt', sep='\t', header=None, index=False)

***
### Negative regulatory elements

Negative regulatory elements (NREs) are called using MACS2 [-log10(P) > 30] comparing input plasmid to STARR-Seq-obtained RNA reads <br> (Model-based Analysis of ChIP-Seq, zhang et al 2008)

 to characterize the effect of enhancers and NREs on expression of their target genes. First, we attempted to assign target genes to CREs on the basis of distance. Each CRE was assigned to the nearest TSS if within 10 kb

In [208]:
nre.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6267 entries, 0 to 6266
Data columns (total 10 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   chromosome      6267 non-null   object 
 1   start_coord     6267 non-null   int64  
 2   end_coord       6267 non-null   int64  
 3   name            6267 non-null   object 
 4   score           6267 non-null   int64  
 5   strand          6267 non-null   object 
 6   signalValue     6267 non-null   float64
 7   pvalue(-log10)  6267 non-null   float64
 8   qvalue(-log10)  6267 non-null   float64
 9   peak            6267 non-null   int64  
dtypes: float64(3), int64(4), object(3)
memory usage: 489.7+ KB


In [209]:
nre.sort_values(by='qvalue(-log10)', ascending=False)

chromosome  start_coord  end_coord  \
5615       chr8     12051697   12052602   
5045       chr6     37786734   37787385   
5887       chr9     41954133   41955150   
5602       chr8      7468847    7469203   
5472       chr7    100887623  100888457   
...         ...          ...        ...   
4005      chr22     44444086   44444330   
1491      chr12    112856426  112856640   
1037      chr11     65625658   65625865   
2018      chr15     73088930   73089178   
1900      chr15     26095865   26096064   

                                                   name  score strand  \
5615  CALLNREreversecontrolTxMACS2STARRENH_indivuall...  39712      .   
5045  CALLNREreversecontrolTxMACS2STARRENH_indivuall...  11355      .   
5887  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   9074      .   
5602  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   8736      .   
5472  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   8231      .   
...                                                 ...    ...    ...   
4005  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    277      .   
1491  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      .   
1037  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      .   
2018  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      .   
1900  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      .   

      signalValue  pvalue(-log10)  qvalue(-log10)  peak  
5615      9.69917      3979.92358      3971.27808   204  
5045      7.55992      1142.65662      1135.56580   514  
5887     24.62481       913.88104       907.41705   679  
5602     42.94287       880.01074       873.62402   149  
5472      4.20734       829.48114       823.15228   960  
...           ...             ...             ...   ...  
4005      7.60395        30.04187        27.70291    58  
1491      1.70155        30.02005        27.68157   241  
1037      1.31368        30.01295        27.67459   123  
2018      4.70075        30.01036        27.67202   728  
1900      2.10375        30.00355        27.66535    13  

[6267 rows x 10 columns]

In [210]:
#Create column with qvalues
nre['qvalue'] = nre['qvalue(-log10)'].map(lambda x: 10 ** -x)

In [211]:
#Drop strand column and rearrange columns
nre = nre[['chromosome', 'start_coord', 'end_coord', 'name', 'score', 'signalValue', 'pvalue(-log10)', 'qvalue(-log10)', 'qvalue', 'peak']]

In [212]:
nre.sort_values(by='qvalue', ascending=False)

chromosome  start_coord  end_coord  \
1900      chr15     26095865   26096064   
2018      chr15     73088930   73089178   
1037      chr11     65625658   65625865   
1491      chr12    112856426  112856640   
4005      chr22     44444086   44444330   
...         ...          ...        ...   
300        chr1    119682919  119683347   
2883      chr19      5720138    5720732   
5615       chr8     12051697   12052602   
2355      chr17       899749     900400   
2868      chr19      3500645    3501017   

                                                   name  score  signalValue  \
1900  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      2.10375   
2018  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      4.70075   
1037  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      1.31368   
1491  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    276      1.70155   
4005  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    277      7.60395   
...                                                 ...    ...          ...   
300   CALLNREreversecontrolTxMACS2STARRENH_indivuall...   8025      3.31381   
2883  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   5997      3.00632   
5615  CALLNREreversecontrolTxMACS2STARRENH_indivuall...  39712      9.69917   
2355  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   3258      2.58787   
2868  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   3428      4.09948   

      pvalue(-log10)  qvalue(-log10)        qvalue  peak  
1900        30.00355        27.66535  2.160976e-28    13  
2018        30.01036        27.67202  2.128041e-28   728  
1037        30.01295        27.67459  2.115485e-28   123  
1491        30.02005        27.68157  2.081757e-28   241  
4005        30.04187        27.70291  1.981938e-28    58  
...              ...             ...           ...   ...  
300        808.86694       802.57648  0.000000e+00   316  
2883       605.46771       599.77106  0.000000e+00   266  
5615      3979.92358      3971.27808  0.000000e+00   204  
2355       330.78641       325.87552  0.000000e+00   354  
2868       347.82938       342.86725  0.000000e+00   254  

[6267 rows x 10 columns]

In [213]:
nre['name'].value_counts()

CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_286479    3
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_292055    2
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_381898    2
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_71365     2
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_138928    2
                                                                                         ..
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_123480    1
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_123412    1
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_123368    1
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_123279    1
CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_

In [214]:
nre[nre['name'].map(nre['name'].value_counts() > 1)]

chromosome  start_coord  end_coord  \
352        chr1    150533611  150533824   
353        chr1    150534234  150534826   
435        chr1    167486799  167487107   
436        chr1    167487691  167487716   
1103      chr11     84431242   84431446   
1104      chr11     84432108   84432343   
1188      chr11    128352963  128353020   
1189      chr11    128353388  128353593   
1260      chr12     14409101   14409276   
1261      chr12     14409467   14409476   
1592      chr13     41634329   41634550   
1593      chr13     41634746   41636161   
1930      chr15     41793940   41794124   
1931      chr15     41794609   41794957   
2341      chr16     88847798   88847904   
2342      chr16     88848081   88848409   
2472      chr17     29034998   29035270   
2473      chr17     29035620   29035937   
3283       chr2     43251777   43251992   
3284       chr2     43252425   43252580   
3634       chr2    237454623  237454865   
3635       chr2    237455233  237455737   
4299       chr3    167812668  167812980   
4300       chr3    167813325  167813797   
4522       chr4    140222520  140223101   
4523       chr4    140223351  140223513   
4593       chr5     14992907   14993188   
4594       chr5     14993402   14993796   
4769       chr5    140892306  140892489   
4770       chr5    140892647  140892996   
4771       chr5    140893643  140893881   
4874       chr5    180644200  180644424   
4875       chr5    180645053  180645416   
5583       chr7    155888765  155889047   
5584       chr7    155889416  155889619   
6113       chrX     40034672   40035149   
6114       chrX     40035498   40035599   
6115       chrX     40430263   40430424   
6116       chrX     40430582   40430782   
6243       chrX    152875748  152875782   
6244       chrX    152875981  152876290   

                                                   name  score  signalValue  \
352   CALLNREreversecontrolTxMACS2STARRENH_indivuall...    282      1.35615   
353   CALLNREreversecontrolTxMACS2STARRENH_indivuall...    282      1.35615   
435   CALLNREreversecontrolTxMACS2STARRENH_indivuall...    477      1.99339   
436   CALLNREreversecontrolTxMACS2STARRENH_indivuall...    477      1.99339   
1103  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   1139     15.50637   
1104  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   1139     15.50637   
1188  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    791      5.08873   
1189  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    791      5.08873   
1260  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    489      2.74288   
1261  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    489      2.74288   
1592  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   3183      4.32224   
1593  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   3183      4.32224   
1930  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    751      3.73819   
1931  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    751      3.73819   
2341  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    716      2.66997   
2342  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    716      2.66997   
2472  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   1291     13.51458   
2473  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   1291     13.51458   
3283  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   1452     26.07928   
3284  CALLNREreversecontrolTxMACS2STARRENH_indivuall...   1452     26.07928   
3634  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    372      2.31731   
3635  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    372      2.31731   
4299  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    340      2.12978   
4300  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    340      2.12978   
4522  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    556      1.45173   
4523  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    556      1.45173   
4593  CALLNREreversecontrolTxMACS2STARRENH_indivuall...    294   

In [215]:
print('There are', len(nre[nre['name'].map(nre['name'].value_counts() > 1)]), 'duplicate names for nre')

There are 41 duplicate names for nre


smallest float is up to e-324

### build 37 to 38

In [216]:
# Drop rows with X and y chromosomes
nre = nre[~nre['chromosome'].isin(['chrX', 'chrY'])]
#cast dataframe to string and create unique names for each row based on position
nre = nre.astype(str)
nre['unique_name'] = nre['chromosome'] +'_'+ nre['start_coord'] +'_'+ nre['end_coord']+ '_' + nre['name']


__Check variable and file name__ (consistent)

In [217]:
#Select relevant columns and write to a txt file for conversion
mpra_nre = nre[['chromosome', 'start_coord', 'end_coord', 'unique_name']]
mpra_nre.to_csv('data/mpra_nre_coords.txt', sep='\t', index=False, header=None)

In [218]:
mpra_nre

chromosome start_coord  end_coord  \
0          chr1      762641     763192   
1          chr1      911493     911786   
2          chr1      948929     949124   
3          chr1      989233     989396   
4          chr1     1152392    1152620   
...         ...         ...        ...   
6063       chr9   139981074  139981542   
6064       chr9   140506301  140506495   
6065       chr9   140512927  140513302   
6066       chr9   140577072  140577228   
6067       chr9   140632813  140632990   

                                            unique_name  
0     chr1_762641_763192_CALLNREreversecontrolTxMACS...  
1     chr1_911493_911786_CALLNREreversecontrolTxMACS...  
2     chr1_948929_949124_CALLNREreversecontrolTxMACS...  
3     chr1_989233_989396_CALLNREreversecontrolTxMACS...  
4     chr1_1152392_1152620_CALLNREreversecontrolTxMA...  
...                                                 ...  
6063  chr9_139981074_139981542_CALLNREreversecontrol...  
6064  chr9_140506301_140506495_CALLNREreversecontrol...  
6065  chr9_140512927_140513302_CALLNREreversecontrol...  
6066  chr9_140577072_140577228_CALLNREreversecontrol...  
6067  chr9_140632813_140632990_CALLNREreversecontrol...  

[6068 rows x 4 columns]

***

### Build 38 conversion inspections

In [219]:
#enh_grch_38 = pd.read_csv('data/enh_grch38.bed', sep='\t', header=None)
enh_grch_38 = pd.read_csv('data/grch38_enh_coords.bed', sep='\t', header=None, names=['chromosome', 'start_coord', 'end_coord', 'name'])

#nre_grch_38 = pd.read_csv('data/nre_grch38.bed', sep='\t', header=None)
nre_grch_38 = pd.read_csv('data/grch38_nre_coords.bed', sep='\t', header=None,names=['chromosome', 'start_coord', 'end_coord', 'name'])

In [220]:
enh_grch_38

chromosome  start_coord  end_coord  \
0          chr1       844682     844790   
1          chr1      1116100    1116462   
2          chr1      1201502    1201726   
3          chr1      1208078    1208173   
4          chr1      1231960    1232156   
...         ...          ...        ...   
8129       chr9    137579026  137579210   
8130       chr9    137590556  137590720   
8131       chr9    137605159  137605312   
8132       chr9    137628626  137628829   
8133       chr9    137722743  137722895   

                                                   name  
0     chr1_780062_780170_MACS2STARRENH_indivuallyrmD...  
1     chr1_1051480_1051842_MACS2STARRENH_indivuallyr...  
2     chr1_1136882_1137106_MACS2STARRENH_indivuallyr...  
3     chr1_1143458_1143553_MACS2STARRENH_indivuallyr...  
4     chr1_1167340_1167536_MACS2STARRENH_indivuallyr...  
...                                                 ...  
8129  chr9_140473478_140473662_MACS2STARRENH_indivua...  
8130  chr9_140485008_140485172_MACS2STARRENH_indivua...  
8131  chr9_140499611_140499764_MACS2STARRENH_indivua...  
8132  chr9_140523078_140523281_MACS2STARRENH_indivua...  
8133  chr9_140617195_140617347_MACS2STARRENH_indivua...  

[8134 rows x 4 columns]

In [221]:
nre_grch_38

chromosome  start_coord  end_coord  \
0          chr1       827261     827812   
1          chr1       976113     976406   
2          chr1      1013549    1013744   
3          chr1      1053853    1054016   
4          chr1      1217012    1217240   
...         ...          ...        ...   
6042       chr9    137086622  137087090   
6043       chr9    137611849  137612043   
6044       chr9    137618475  137618850   
6045       chr9    137682620  137682776   
6046       chr9    137738361  137738538   

                                                   name  
0     chr1_762641_763192_CALLNREreversecontrolTxMACS...  
1     chr1_911493_911786_CALLNREreversecontrolTxMACS...  
2     chr1_948929_949124_CALLNREreversecontrolTxMACS...  
3     chr1_989233_989396_CALLNREreversecontrolTxMACS...  
4     chr1_1152392_1152620_CALLNREreversecontrolTxMA...  
...                                                 ...  
6042  chr9_139981074_139981542_CALLNREreversecontrol...  
6043  chr9_140506301_140506495_CALLNREreversecontrol...  
6044  chr9_140512927_140513302_CALLNREreversecontrol...  
6045  chr9_140577072_140577228_CALLNREreversecontrol...  
6046  chr9_140632813_140632990_CALLNREreversecontrol...  

[6047 rows x 4 columns]

In [222]:
len(mpra_nre) - len(nre_grch_38)

21

In [223]:
len(enh_bed) - len(enh_grch_38)

21

Check which names are missing for the nre dataset after converting

In [224]:
missing_nre = mpra_nre[~mpra_nre['unique_name'].isin(nre_grch_38['name'])]
missing_nre

chromosome start_coord  end_coord  \
306        chr1   142890760  142891321   
307        chr1   143283666  143283895   
332        chr1   147624310  147624672   
335        chr1   147931299  147931653   
338        chr1   148598378  148598910   
342        chr1   149514459  149514637   
343        chr1   149575248  149575961   
345        chr1   149605660  149605874   
346        chr1   149672222  149672696   
696       chr10    48828008   48828181   
700       chr10    51503121   51503287   
1884      chr14   106552119  106552392   
1892      chr15    22833123   22833546   
1894      chr15    23482232   23482426   
2225      chr16    46391134   46391693   
2226      chr16    46399481   46399677   
2227      chr16    46404914   46405368   
2228      chr16    46406802   46407088   
2489      chr17    34091493   34091784   
2490      chr17    34207265   34207442   
2491      chr17    34251908   34252065   
2492      chr17    34430966   34431336   
2493      chr17    34957704   34958331   
2494      chr17    35504308   35504545   
2495      chr17    36442474   36442844   
2496      chr17    36452737   36453351   
2497      chr17    36858715   36859016   
3051      chr19    40404125   40404570   
5885       chr9    40791887   40792202   
5889       chr9    45439576   45439922   
5890       chr9    45441853   45442246   
5891       chr9    45729204   45729728   
5895       chr9    68799381   68799893   
5896       chr9    69137532   69137747   

                                            unique_name  
306   chr1_142890760_142891321_CALLNREreversecontrol...  
307   chr1_143283666_143283895_CALLNREreversecontrol...  
332   chr1_147624310_147624672_CALLNREreversecontrol...  
335   chr1_147931299_147931653_CALLNREreversecontrol...  
338   chr1_148598378_148598910_CALLNREreversecontrol...  
342   chr1_149514459_149514637_CALLNREreversecontrol...  
343   chr1_149575248_149575961_CALLNREreversecontrol...  
345   chr1_149605660_149605874_CALLNREreversecontrol...  
346   chr1_149672222_149672696_CALLNREreversecontrol...  
696   chr10_48828008_48828181_CALLNREreversecontrolT...  
700   chr10_51503121_51503287_CALLNREreversecontrolT...  
1884  chr14_106552119_106552392_CALLNREreversecontro...  
1892  chr15_22833123_22833546_CALLNREreversecontrolT...  
1894  chr15_23482232_23482426_CALLNREreversecontrolT...  
2225  chr16_46391134_46391693_CALLNREreversecontrolT...  
2226  chr16_46399481_46399677_CALLNREreversecontrolT...  
2227  chr16_46404914_46405368_CALLNREreversecontrolT...  
2228  chr16_46406802_46407088_CALLNREreversecontrolT...  
2489  chr17_34091493_34091784_CALLNREreversecontrolT...  
2490  chr17_34207265_34207442_CALLNREreversecontrolT...  
2491  chr17_34251908_34252065_CALLNREreversecontrolT...  
2492  chr17_34430966_34431336_CALLNREreversecontrolT...  
2493  chr17_34957704_34958331_CALLNREreversecontrolT...  
2494  chr17_35504308_35504545_CALLNREreversecontrolT...  
2495  chr17_36442474_36442844_CALLNREreversecontrolT...  
2496  chr17_36452737_36453351_CALLNREreversecontrolT...  
2497  chr17_36858715_36859016_CALLNREreversecontrolT...  
3051  chr19_40404125_40404570_CALLNREreversecontrolT...  
5885  chr9_40791887_40792202_CALLNREreversecontrolTx...  
5889  chr9_45439576_45439922_CALLNREreversecontrolTx...  
5890  chr9_45441853_45442246_CALLNREreversecontrolTx...  
5891  chr9_45729204_45729728_CALLNREreversecontrolTx...  
5895  chr9_68799381_68799893_CALLNREreversecontrolTx...  
5896  chr9_69137532_69137747_CALLNREreversecontrolTx...

In [225]:
print(f'There are {len(missing_nre)} entries/names that are not converted')

There are 34 entries/names that are not converted


The difference before and after converting is however 21, so there are some entries added.

Example chromosome 20

In [226]:
nre_chr20 = mpra_nre[mpra_nre['chromosome'] == 'chr20']
nre_chr20

chromosome start_coord end_coord  \
3649      chr20      311259    311448   
3650      chr20      442443    442721   
3651      chr20      524388    524786   
3652      chr20     1305786   1306425   
3653      chr20     1878619   1878804   
...         ...         ...       ...   
3800      chr20    62289569  62289973   
3801      chr20    62368729  62368950   
3802      chr20    62435417  62435584   
3803      chr20    62612178  62612680   
3804      chr20    62917281  62917599   

                                            unique_name  
3649  chr20_311259_311448_CALLNREreversecontrolTxMAC...  
3650  chr20_442443_442721_CALLNREreversecontrolTxMAC...  
3651  chr20_524388_524786_CALLNREreversecontrolTxMAC...  
3652  chr20_1305786_1306425_CALLNREreversecontrolTxM...  
3653  chr20_1878619_1878804_CALLNREreversecontrolTxM...  
...                                                 ...  
3800  chr20_62289569_62289973_CALLNREreversecontrolT...  
3801  chr20_62368729_62368950_CALLNREreversecontrolT...  
3802  chr20_62435417_62435584_CALLNREreversecontrolT...  
3803  chr20_62612178_62612680_CALLNREreversecontrolT...  
3804  chr20_62917281_62917599_CALLNREreversecontrolT...  

[156 rows x 4 columns]

In [227]:
nre_chr20_38 = nre_grch_38[nre_grch_38['chromosome'] == 'chr20']
nre_chr20_38

chromosome  start_coord  end_coord  \
3628      chr20       330615     330804   
3629      chr20       461799     462077   
3630      chr20       543744     544142   
3631      chr20      1325142    1325781   
3632      chr20      1897973    1898158   
...         ...          ...        ...   
3780      chr20     63658216   63658620   
3781      chr20     63737376   63737597   
3782      chr20     63804064   63804231   
3783      chr20     63980825   63981327   
3784      chr20     64285928   64286246   

                                                   name  
3628  chr20_311259_311448_CALLNREreversecontrolTxMAC...  
3629  chr20_442443_442721_CALLNREreversecontrolTxMAC...  
3630  chr20_524388_524786_CALLNREreversecontrolTxMAC...  
3631  chr20_1305786_1306425_CALLNREreversecontrolTxM...  
3632  chr20_1878619_1878804_CALLNREreversecontrolTxM...  
...                                                 ...  
3780  chr20_62289569_62289973_CALLNREreversecontrolT...  
3781  chr20_62368729_62368950_CALLNREreversecontrolT...  
3782  chr20_62435417_62435584_CALLNREreversecontrolT...  
3783  chr20_62612178_62612680_CALLNREreversecontrolT...  
3784  chr20_62917281_62917599_CALLNREreversecontrolT...  

[157 rows x 4 columns]

In [228]:
# Check for duplicate names
nre_chr20_38['name'].value_counts()

chr20_34042543_34043186_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_203530    2
chr20_44002291_44002529_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_205089    1
chr20_44452811_44453123_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_205143    1
chr20_44600177_44601025_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_205176    1
chr20_45318028_45318410_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_205286    1
                                                                                                                 ..
chr20_30606195_30606375_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_202916    1
chr20_30619231_30619455_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_202918    1
chr20_30697273_30697794_CALLNREreversecontrolTxMACS2STARRENH_indivuallyr

Inspect the duplicate name

In [229]:
mpra_nre[mpra_nre['unique_name'] == 'chr20_34042543_34043186_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_203530']

chromosome start_coord end_coord  \
3717      chr20    34042543  34043186   

                                            unique_name  
3717  chr20_34042543_34043186_CALLNREreversecontrolT...

In [230]:
nre_grch_38[nre_grch_38['name']== 'chr20_34042543_34043186_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_203530']

chromosome  start_coord  end_coord  \
3696      chr20     35454721   35455117   
3697      chr20     35455118   35455365   

                                                   name  
3696  chr20_34042543_34043186_CALLNREreversecontrolT...  
3697  chr20_34042543_34043186_CALLNREreversecontrolT...

One coordinate from build 37 can give multiple coordinates for build 38 <br>
Check for all duplicate entries:

In [231]:
#Check all duplicate entries after converting
nre_grch_38[nre_grch_38['name'].map(nre_grch_38['name'].value_counts() >1)]

chromosome  start_coord  end_coord  \
311        chr1    149176821  149177115   
312        chr1    149176650  149176820   
313        chr1    149102832  149102980   
314        chr1    149102759  149102808   
333        chr1    148303201  148303374   
334        chr1    148303374  148303376   
335        chr1    148303386  148303630   
2436      chr17     22377474   22377500   
2437      chr17     22377518   22377575   
2438      chr17     22377575   22377650   
2632      chr17     74862145   74862370   
2633      chr17     74862371   74862454   
3696      chr20     35454721   35455117   
3697      chr20     35455118   35455365   
5390       chr7     73324346   73324351   
5391       chr7     73324352   73324590   
5398       chr7     74890318   74890595   
5399       chr7     74890597   74891002   
5401       chr7     75359107   75359200   
5402       chr7     75359209   75359321   
5403       chr7     75359321   75359596   
5795       chr8    144365995  144366102   
5796       chr8    144366102  144366153   

                                                   name  
311   chr1_144519839_144520373_CALLNREreversecontrol...  
312   chr1_144519839_144520373_CALLNREreversecontrol...  
313   chr1_144594255_144594452_CALLNREreversecontrol...  
314   chr1_144594255_144594452_CALLNREreversecontrol...  
333   chr1_147775312_147775737_CALLNREreversecontrol...  
334   chr1_147775312_147775737_CALLNREreversecontrol...  
335   chr1_147775312_147775737_CALLNREreversecontrol...  
2436  chr17_21904441_21904600_CALLNREreversecontrolT...  
2437  chr17_21904441_21904600_CALLNREreversecontrolT...  
2438  chr17_21904441_21904600_CALLNREreversecontrolT...  
2632  chr17_72858268_72858576_CALLNREreversecontrolT...  
2633  chr17_72858268_72858576_CALLNREreversecontrolT...  
3696  chr20_34042543_34043186_CALLNREreversecontrolT...  
3697  chr20_34042543_34043186_CALLNREreversecontrolT...  
5390  chr7_72738345_72738588_CALLNREreversecontrolTx...  
5391  chr7_72738345_72738588_CALLNREreversecontrolTx...  
5398  chr7_74306439_74307121_CALLNREreversecontrolTx...  
5399  chr7_74306439_74307121_CALLNREreversecontrolTx...  
5401  chr7_74988361_74988865_CALLNREreversecontrolTx...  
5402  chr7_74988361_74988865_CALLNREreversecontrolTx...  
5403  chr7_74988361_74988865_CALLNREreversecontrolTx...  
5795  chr8_145589655_145589814_CALLNREreversecontrol...  
5796  chr8_145589655_145589814_CALLNREreversecontrol...

In [233]:
duplicates_nre = nre_grch_38[nre_grch_38['name'].duplicated()]
duplicates_nre = list(duplicates_nre['name'].unique())
print(f'There are {len(duplicates_nre)} unique duplicate entrys/names')
print(f'Corresponding to {len(nre_grch_38[nre_grch_38.name.isin(duplicates_nre)])} rows')

There are 10 unique duplicate entrys/names
Corresponding to 23 rows


In [234]:
duplicates_nre

['chr1_144519839_144520373_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_17689',
 'chr1_144594255_144594452_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_17701',
 'chr1_147775312_147775737_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_18075',
 'chr17_21904441_21904600_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_138276',
 'chr17_72858268_72858576_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_145627',
 'chr20_34042543_34043186_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_203530',
 'chr7_72738345_72738588_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_324915',
 'chr7_74306439_74307121_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_325207',
 'chr7_74988361_74988865_CALLNREreversecontrolTxMACS2STARRENH_indivuallyrmDupat75t

In [235]:
#drop rows with a duplicate entry
nre_grch_38 = nre_grch_38.drop_duplicates(subset='name', keep=False)

In [238]:
nre_grch_38 #6047

chromosome  start_coord  end_coord  \
0          chr1       827261     827812   
1          chr1       976113     976406   
2          chr1      1013549    1013744   
3          chr1      1053853    1054016   
4          chr1      1217012    1217240   
...         ...          ...        ...   
6042       chr9    137086622  137087090   
6043       chr9    137611849  137612043   
6044       chr9    137618475  137618850   
6045       chr9    137682620  137682776   
6046       chr9    137738361  137738538   

                                                   name  
0     chr1_762641_763192_CALLNREreversecontrolTxMACS...  
1     chr1_911493_911786_CALLNREreversecontrolTxMACS...  
2     chr1_948929_949124_CALLNREreversecontrolTxMACS...  
3     chr1_989233_989396_CALLNREreversecontrolTxMACS...  
4     chr1_1152392_1152620_CALLNREreversecontrolTxMA...  
...                                                 ...  
6042  chr9_139981074_139981542_CALLNREreversecontrol...  
6043  chr9_140506301_140506495_CALLNREreversecontrol...  
6044  chr9_140512927_140513302_CALLNREreversecontrol...  
6045  chr9_140577072_140577228_CALLNREreversecontrol...  
6046  chr9_140632813_140632990_CALLNREreversecontrol...  

[6024 rows x 4 columns]

__enhancer grch38 inspection__

In [239]:
missing_enh = enh_bed[~enh_bed['unique_name'].isin(enh_grch_38['name'])]
missing_enh

chromosome start_coord  end_coord  \
460        chr1   145096358  145096495   
484        chr1   149513997  149514141   
485        chr1   149605477  149605586   
1012      chr10    48477923   48478051   
1013      chr10    48481322   48481507   
1015      chr10    51565073   51565230   
2620      chr14   106551152  106551289   
2630      chr15    23115604   23115656   
3046      chr16    46395403   46395438   
3047      chr16    46406769   46406867   
3343      chr17    34136375   34136492   
3344      chr17    34217772   34217983   
3345      chr17    34842334   34842517   
3346      chr17    34985853   34985959   
3347      chr17    35447080   35447257   
3348      chr17    35969406   35969688   
3349      chr17    36884601   36884741   
3350      chr17    36886289   36886530   
3351      chr17    36908717   36909023   
3352      chr17    36981520   36981686   
7906       chr9    45357059   45357206   
7908       chr9    68726378   68726478   
7909       chr9    69065426   69065573   

                                            unique_name  
460   chr1_145096358_145096495_MACS2STARRENH_indivua...  
484   chr1_149513997_149514141_MACS2STARRENH_indivua...  
485   chr1_149605477_149605586_MACS2STARRENH_indivua...  
1012  chr10_48477923_48478051_MACS2STARRENH_indivual...  
1013  chr10_48481322_48481507_MACS2STARRENH_indivual...  
1015  chr10_51565073_51565230_MACS2STARRENH_indivual...  
2620  chr14_106551152_106551289_MACS2STARRENH_indivu...  
2630  chr15_23115604_23115656_MACS2STARRENH_indivual...  
3046  chr16_46395403_46395438_MACS2STARRENH_indivual...  
3047  chr16_46406769_46406867_MACS2STARRENH_indivual...  
3343  chr17_34136375_34136492_MACS2STARRENH_indivual...  
3344  chr17_34217772_34217983_MACS2STARRENH_indivual...  
3345  chr17_34842334_34842517_MACS2STARRENH_indivual...  
3346  chr17_34985853_34985959_MACS2STARRENH_indivual...  
3347  chr17_35447080_35447257_MACS2STARRENH_indivual...  
3348  chr17_35969406_35969688_MACS2STARRENH_indivual...  
3349  chr17_36884601_36884741_MACS2STARRENH_indivual...  
3350  chr17_36886289_36886530_MACS2STARRENH_indivual...  
3351  chr17_36908717_36909023_MACS2STARRENH_indivual...  
3352  chr17_36981520_36981686_MACS2STARRENH_indivual...  
7906  chr9_45357059_45357206_MACS2STARRENH_indivuall...  
7908  chr9_68726378_68726478_MACS2STARRENH_indivuall...  
7909  chr9_69065426_69065573_MACS2STARRENH_indivuall...

In [240]:
print(f'There are {len(missing_enh)} entries/names that are not converted')

There are 23 entries/names that are not converted


In [247]:
enh_grch_38['name'].value_counts()

chr1_145454616_145454878_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_9891      2
chr14_106995875_106996150_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_62780    2
chr1_780062_780170_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_7               1
chr3_45883605_45883739_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_124297      1
chr3_45730624_45730793_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_124285      1
                                                                                           ..
chr15_52774217_52774391_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_64981      1
chr15_52338124_52338362_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_64933      1
chr15_52029846_52030175_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_64901      1
chr15_51057627_51057705_MACS2STARRENH_indivuallyrmDupat75thpercentile021623_peak_64830      1
chr9_140617195_140617347_MACS2STARRENH_indivuallyrmDupat75th

In [250]:
#Check all duplicate entries after converting
enh_grch_38[enh_grch_38['name'].map(enh_grch_38['name'].value_counts() >1)]

chromosome  start_coord  end_coord  \
462        chr1    145980374  145980466   
463        chr1    145980205  145980374   
2618      chr14    106539879  106539950   
2619      chr14    106539951  106540155   

                                                   name  
462   chr1_145454616_145454878_MACS2STARRENH_indivua...  
463   chr1_145454616_145454878_MACS2STARRENH_indivua...  
2618  chr14_106995875_106996150_MACS2STARRENH_indivu...  
2619  chr14_106995875_106996150_MACS2STARRENH_indivu...

In [251]:
duplicates_enh = enh_grch_38[enh_grch_38['name'].duplicated()]
duplicates_enh = list(duplicates_enh['name'].unique())
print(f'There are {len(duplicates_enh)} unique duplicate entrys/names')
print(f'Corresponding to {len(enh_grch_38[enh_grch_38.name.isin(duplicates_enh)])} rows')

There are 2 unique duplicate entrys/names
Corresponding to 4 rows


In [252]:
enh_grch_38 = enh_grch_38.drop_duplicates(subset='name', keep=False)

In [253]:
enh_grch_38

chromosome  start_coord  end_coord  \
0          chr1       844682     844790   
1          chr1      1116100    1116462   
2          chr1      1201502    1201726   
3          chr1      1208078    1208173   
4          chr1      1231960    1232156   
...         ...          ...        ...   
8129       chr9    137579026  137579210   
8130       chr9    137590556  137590720   
8131       chr9    137605159  137605312   
8132       chr9    137628626  137628829   
8133       chr9    137722743  137722895   

                                                   name  
0     chr1_780062_780170_MACS2STARRENH_indivuallyrmD...  
1     chr1_1051480_1051842_MACS2STARRENH_indivuallyr...  
2     chr1_1136882_1137106_MACS2STARRENH_indivuallyr...  
3     chr1_1143458_1143553_MACS2STARRENH_indivuallyr...  
4     chr1_1167340_1167536_MACS2STARRENH_indivuallyr...  
...                                                 ...  
8129  chr9_140473478_140473662_MACS2STARRENH_indivua...  
8130  chr9_140485008_140485172_MACS2STARRENH_indivua...  
8131  chr9_140499611_140499764_MACS2STARRENH_indivua...  
8132  chr9_140523078_140523281_MACS2STARRENH_indivua...  
8133  chr9_140617195_140617347_MACS2STARRENH_indivua...  

[8130 rows x 4 columns]

Write the dataframes to new txt files

In [76]:
enh_grch_38.to_csv('data/enh_cd4_coords.txt', sep='\t', index=False, header=None)
nre_grch_38.to_csv('data/nre_cd4_coords.txt', sep='\t', index=False, header=None)

***

Overlap analysis

In [ ]:
header_enh = ['chromosome_enh', 'start_enh', 'end_enh', 'enh_name','chromosome_lcl', 'start_lcl', 'end_lcl', 'snp']
header_nre = ['chromosome_nre', 'start_nre', 'end_nre', 'nre_name','chromosome_lcl', 'start_lcl', 'end_lcl', 'snp']

In [ ]:
enh_overlap_nre = pd.read_csv('data/enh_overlap_with_nre.txt', sep='\t', header=None)
enh_overlap_sig_lcl = pd.read_csv('data/enh_overlap_sig_lcl.txt', sep='\t', header=None, names=header_enh)
enh_overlap_total_lcl = pd.read_csv('data/enh_overlap_total_lcl.txt', sep='\t', header=None, names=header_enh)
nre_overlap_sig_lcl = pd.read_csv('data/nre_overlap_sig_lcl.txt', sep='\t', header=None, names=header_nre)
nre_overlap_total_lcl = pd.read_csv('data/nre_overlap_total_lcl.txt', sep='\t', header=None, names=header_nre)

In [ ]:
enh_overlap_sig_lcl

chromosome_enh  start_enh    end_enh  \
0             chr1    2546978    2547299   
1             chr1   22025961   22026209   
2             chr1   26169775   26169929   
3             chr1   27914537   27914833   
4             chr1   43389750   43389967   
..             ...        ...        ...   
108           chr9  113401339  113401554   
109           chr9  128275996  128276234   
110           chr9  136118927  136119153   
111           chr9  136403654  136403775   
112           chr9  136403654  136403775   

                                              enh_name chromosome_lcl  \
0    chr1_2478417_2478738_MACS2STARRENH_indivuallyr...           chr1   
1    chr1_22352454_22352702_MACS2STARRENH_indivuall...           chr1   
2    chr1_26496266_26496420_MACS2STARRENH_indivuall...           chr1   
3    chr1_28241048_28241344_MACS2STARRENH_indivuall...           chr1   
4    chr1_43855421_43855638_MACS2STARRENH_indivuall...           chr1   
..                                                 ...            ...   
108  chr9_116163619_116163834_MACS2STARRENH_indivua...           chr9   
109  chr9_131038275_131038513_MACS2STARRENH_indivua...           chr9   
110  chr9_139010773_139010999_MACS2STARRENH_indivua...           chr9   
111  chr9_139298106_139298227_MACS2STARRENH_indivua...           chr9   
112  chr9_139298106_139298227_MACS2STARRENH_indivua...           chr9   

     start_lcl    end_lcl              snp  
0      2546997    2547147        rs2477678  
1     22025827   22025977  chr1:22352395:D  
2     26169804   26169954        rs2232646  
3     27914497   27914647       rs17185052  
4     43389832   43389982         rs839753  
..         ...        ...              ...  
108  113401290  113401440        rs1771220  
109  128276221  128276371      rs115231472  
110  136118794  136118944       rs79362369  
111  136403686  136403836       rs80029103  
112  136403698  136403848       rs75859952  

[113 rows x 8 columns]

In [ ]:
enh_overlap_total_lcl

chromosome_enh  start_enh    end_enh  \
0             chr1    2546978    2547299   
1             chr1   22025961   22026209   
2             chr1   26169775   26169929   
3             chr1   27914537   27914833   
4             chr1   43389750   43389967   
..             ...        ...        ...   
161           chr9  113401339  113401554   
162           chr9  128275996  128276234   
163           chr9  136118927  136119153   
164           chr9  136403654  136403775   
165           chr9  136403654  136403775   

                                              enh_name chromosome_lcl  \
0    chr1_2478417_2478738_MACS2STARRENH_indivuallyr...           chr1   
1    chr1_22352454_22352702_MACS2STARRENH_indivuall...           chr1   
2    chr1_26496266_26496420_MACS2STARRENH_indivuall...           chr1   
3    chr1_28241048_28241344_MACS2STARRENH_indivuall...           chr1   
4    chr1_43855421_43855638_MACS2STARRENH_indivuall...           chr1   
..                                                 ...            ...   
161  chr9_116163619_116163834_MACS2STARRENH_indivua...           chr9   
162  chr9_131038275_131038513_MACS2STARRENH_indivua...           chr9   
163  chr9_139010773_139010999_MACS2STARRENH_indivua...           chr9   
164  chr9_139298106_139298227_MACS2STARRENH_indivua...           chr9   
165  chr9_139298106_139298227_MACS2STARRENH_indivua...           chr9   

     start_lcl    end_lcl              snp  
0      2546997    2547147        rs2477678  
1     22025827   22025977  chr1:22352395:D  
2     26169804   26169954        rs2232646  
3     27914497   27914647       rs17185052  
4     43389832   43389982         rs839753  
..         ...        ...              ...  
161  113401290  113401440        rs1771220  
162  128276221  128276371      rs115231472  
163  136118794  136118944       rs79362369  
164  136403686  136403836       rs80029103  
165  136403698  136403848       rs75859952  

[166 rows x 8 columns]

In [ ]:
f'Unique enhancer (starr) coordinates that overlap with significant lcl mpra coordinates: {len(enh_overlap_sig_lcl.enh_name.value_counts())}'

'Unique enhancer (starr) coordinates that overlap with significant lcl mpra coordinates: 99'

In [ ]:
f'Unique enhancer(starr) coordinates that overlap with all lcl mpra coordinates: {len(enh_overlap_total_lcl.enh_name.value_counts())}'

'Unique enhancer (starr) coordinates that overlap with lcl mpra: 138'

In [ ]:
#Check the names of enhancer coordinates that are not in the overlap with significant lcl mpra's dataframe
not_overlapping = enh_overlap_total_lcl[~enh_overlap_total_lcl.enh_name.isin(enh_overlap_sig_lcl.enh_name)]
f'Unique enhancer coordinates that overlap with total and not significant lcl mpra coordinates: {len(not_overlapping.enh_name.value_counts())}'

'Unique enhancer coordinates that overlap with total and not significant lcl mpra coordinates: 39'

In [ ]:
#Check the names of enhancer coordinates that are not in the overlap with significant lcl mpra's dataframe
not_overlapping = enh_overlap_total_lcl[~enh_overlap_total_lcl.snp.isin(enh_overlap_sig_lcl.snp)]
f'Unique snps that are in total and not significant lcl snps: {len(not_overlapping.snp.value_counts())}'

'Unique snps that are in total and not significant lcl snps: 53'

- enhancer - sig lcl mpra: 99 coordinates overlap
- enhancer - total lcl mpra: 138 coordinates overlap
- difference in overlap coordinates: 39

In [ ]:
f'Unique nre (starr) coordinates that overlap with significant lcl mpra coordinates: {len(nre_overlap_sig_lcl.nre_name.value_counts())}'


'Unique enhancer (starr) coordinates that overlap with significant lcl mpra coordinates: 79'

In [ ]:
f'Unique enhancer(starr) coordinates that overlap with all lcl mpra coordinates: {len(nre_overlap_total_lcl.nre_name.value_counts())}'

'Unique enhancer(starr) coordinates that overlap with all lcl mpra coordinates: 137'

In [ ]:
#Check the names of enhancer coordinates that are not in the overlap with significant lcl mpra's dataframe
not_overlapping_nre = nre_overlap_total_lcl[~nre_overlap_total_lcl.nre_name.isin(nre_overlap_sig_lcl.nre_name)]
f'Unique enhancer coordinates that overlap with total and not significant lcl mpra coordinates: {len(not_overlapping_nre.nre_name.value_counts())}'

'Unique enhancer coordinates that overlap with total and not significant lcl mpra coordinates: 58'

Enhancers were called using MACS2 in which the plasmid reads were used as a control, [macs2 callpeak -f BEDPE –keep-dup all -g 2.7E9], and with -log10(P-value) greater than 75. NRE were also detected with MACS2, using the plasmid reads as treatment and the STARR RNA as control [macs2 callpeak -f BEDPE –keep-dup all -g 2.7E9], with -log10(P-value) greater than 30. Largely similar results can be obtained using FAST-NR [77]. Both enhancers and NRE are retained only if intersecting an ATAC-Seq narrowPeak called using MACS2 from the plasmid reads alone.

All final values and CRE locations are listed in Supplemental Table 2. Detailed data processing script can be found at https://github.com/Barski-lab/Lenti-STARR

In [171]:
len(enh_grch_38)

8130